### ISSUE TITLE
Lazy dynamic imports

### ISSUE THREAD

**Issue by IlyaSemenov** [unjs/nitro#1670]
### Describe the feature

As a developer, I would like to be able to import some modules on demand, making sure that they are not loaded until I actually need them (in particular, making sure that they cause no side effects until `import(".../module")` is actually called).

Currently, that is not so.

For example, given a Nuxt 3.7 project, if I create an API handler at `server/api/test.ts`:

```ts
// The function is never called
async function test() {
  console.log("Entered test")
  await import("../bar")
}

console.log("test is of type", typeof test)
```

and `server/bar.ts`:

```ts
throw new Error("I should never be called.")

export default function () {}
```

and run `nuxi dev`, the bar module will be imported without `import()` actually ever called:

```
> nuxi dev

[9:19:34 AM] Nuxt 3.7.0 with Nitro 2.6.2
[9:19:34 AM] 
  ➜ Local:    http://localhost:3000/
  ➜ Network:  use --host to expose

[9:19:36 AM] ℹ Vite client warmed up in 602ms
[nitro 9:19:37 AM] ✔ Nitro built in 347 ms
test is of type function
[nitro] [uncaughtException] Error: I should never be called.
    at <anonymous> (/Users/semenov/nuxt-test/server/bar.ts:1:7)
    at ModuleJob.run (node:internal/modules/esm/module_job:192:25)
    at async DefaultModuleLoader.import (node:internal/modules/esm/loader:228:24)
    at async loadESM (node:internal/process/esm_loader:40:7)
    at async handleMainPromise (node:internal/modules/run_main:66:12)
```

This is **not** how `import` works in plain ESM or [webpack](https://webpack.js.org/api/module-methods/#webpackmode) or vite, which all use lazy import by default.

The use case for that is I want to refer to a collection of modules with side effects (database migrations) and import them programmatically as needed. With how currently nitro works, they are all imported at once on app start, regardless of whether my code actually calls `import()` or not.

Related user-land issue: https://github.com/romeerez/orchid-orm/issues/167

### Additional information

- [ ] Would you be willing to help implement this feature?

**Comment by markthree** [comment]
@IlyaSemenov Hi, you can use the following configuration to achieve results
```ts
// nitro.config.ts
export default defineNitroConfig({
  inlineDynamicImports: false,
});
```

or `nuxt` project

```ts
// nuxt.config.ts
export default defineNuxtConfig({
  nitro: {
    inlineDynamicImports: false,
  }
});
```

**Comment by pi0** [comment]
@markthree `inlineDynamicImports` flag is disabled for some presets only and i wouldn't recommend to modify the defaults

@IlyaSemenov Please consider that, while Nitro uses dynamic imports for performance reasons, you should never rely on this fact to introduce side-effects in event handlers. Based on target bundling requirements this behavior can change and we might even introduce eager chunk warm-ups.

The safest method is to leverage `defineLazyEventHandler`:

```
export default lazyEventHandler(() => {
  // Lazy logic here
  return eventHandler((event) => {
    
  })
})
```

**Comment by markthree** [comment]
@pi0 Yes, I'm sorry I didn't explain the reason.

@IlyaSemenov On some cloud services, dynamic imports is disabled for security reasons. So when setting inlineDynamicImports, please check if your deployment service supports dynamic imports。

**Comment by IlyaSemenov** [comment]
Thank you, `inlineDynamicImports: false` fixes it, indeed.

I still however not following why the lazy imports can be somehow not supported. Aren't they transpiled? I mean, I can build a bundle with vite or esbuild and it will work on any deployment service regardless of whether the source was using dynamic (lazy) imports, as there won't be the actual ESM `import` calls (but rather some `__import` functions which simply call a bundled function — the deployment service can't possibly know these were modules before bundling).

If the imports are not transpiled by nitro into function calls, that explains it, but may be it means that instead of `inlineDynamicImports: true` it should be rather `transpileDynamicImports: true`? Granted, that's probably a whole lot of work, and of course I'm not "demanding" it...just curious about these decisions.

**Comment by pi0** [comment]
Please consider not customizing `inlineDynamicImports` it is not a solution and can cause you other issues.

Lazy imports are supported, and used in production builds whenever possible. We might enable it by default for dev server experience in the future as well.

However, having side effects in modules in general is not a good practice. It makes the whole server lifecycle unpredictable.

**Comment by IlyaSemenov** [comment]
I totally agree, this is a bad practice in general. But sometimes we have third party libraries, so that would be better if there was a way to handle that this way or another. Anyhow, thank you for the responses.

**Comment by cjpearson** [comment]
@pi0 do you know if there's a proper solution for dealing with third-party libraries? 

I'm working on a Nuxt module which depends on unstorage and should work in node and edge environments. The redis driver can be used for apps running in node environments, but another driver should be used for edge runtimes. Unfortunately even a dynamic import of `unstorage/drivers/redis` executes code on startup that will fail in a worker environment.

`inlineDynamicImports: false` works, but for the reasons mentioned above it would be nice to avoid it.

**Comment by pi0** [comment]
@cjpearson You can easily configure nitro storage differently based on the deployment in your module so that it only includes required driver. Feel free to make an issue and ping me in your module if needed any help 👍🏼

**Comment by cjpearson** [comment]
Hi @pi0 thanks for the response! Unfortunately the code isn't public at this point, but looking again at the nitro storage [docs](https://nitro.unjs.io/guide/storage), I think may have misunderstood how to best implement this.

The current approach is to pass all the storage options to the module, something like:

```
myModule: {
    storage: {
        provider: 'redis',
        options: {...}
    }
}
```

and then the module's handler would initialize the storage driver. 

But now I think the correct method is to have the app define a storage mountpoint (say 'myModule') in the nitro options (or have the module do it in setup, based on the module config). The module's handler would then use something like `useStorage('myModule')` instead of calling `createStorage`. Does that make sense?

**Comment by pi0** [comment]
Yes exactly you need to define it for nitro storage options and access from useStorage 👍🏼

**Comment by cjpearson** [comment]
Ok, thanks for the clarification!
